### ISSUE TITLE
Remove implicit "function" in units, instead make them explicit

### ISSUE THREAD

**Issue by fingolfin** [fingolfin/UnitTests#1]
Currently, .unit tests files are implicitly wrapped into a function. That means they cannot be read as GAP files. Øystein suggested that this be changed: By making the function declaration explicit, it is much easier for the user to understand that goes on. Also, it is possible to declare multiple test suites in a single file. Finally, this would make it possible to give explicit parameter lists to the functions wrapping the test suites, which could be used for error checking.

Here is how that could look like:

```
InstallTest( function()
  AddFailingTestCase("some_expected_fail", function()
      AssertEqual( 42, 7 );
  end);
end); # InstallTest

InstallTest( function(a, b, c)
  AddTestCase("add_associative", function()
      AssertEqual( (a + b) + c, a + (b + c) );
  end);
end); # InstallTest
```

**Comment by fingolfin** [comment]
Another potential syntax variant:

```
InstallTest( function( suite )
    local G, oldlevel;

# Optionally perform some setup whenever an instance
# of this test suite is run.
suite.setup := function()
    oldlevel := InfoLevel(InfoWarning);
    SetInfoLevel(InfoWarning, 1);
    G := SmallGroup(8, 4);
    return true; # return false to indicate a setup failure
end);

# Optionally perform some setup after an instance
# of this test suite has been run.
suite.teardown := function()
    SetInfoLevel(InfoWarning, oldlevel);
end;

# a working test
suite.tests.group_size := function()
    AssertEqual( Size(G), 8 );
    AssertCmp( Size(G), "<", 9  );
    AssertTrue( Size(G) = 8 );
end;

# a known-to-fail test
suite.xfail.size_expected_fail := function()
    AssertEqual( Size(G), 7 );
end;

end); # InstallTest
```

**Comment by fingolfin** [comment]
(Some inspiration can e.g. be taken from https://docs.python.org/3.5/library/unittest.html)

**Comment by oysteins** [comment]
I think both these alternative syntaxes seem fine.

Some related suggestions:
- `InstallTest` could take a filter list as a separate argument, so that parametrized test suites can easily check the types of the arguments they receive.
- Maybe `InstallTest` should be named `InstallTestSuite`, to make it clearer that it installs a test suite and not a test case.

**Comment by fingolfin** [comment]
Indeed. Also, perhaps the package name should also be a parameter... 

One other idea: The funtion `InstallTestSuite` (or perhaps `InstallUnitTests`, or...?) should return an object representing the created suite, so that one can manipulate it later on (e.g. instantiate it with various parameters)

**Comment by oysteins** [comment]
It seems reasonable to have the package name stored in the test suites, but is it necessary to explicitly pass it as a parameter when creating a test suite?  Assuming that the tests are stored inside the package's directory, would it be possible for the testing framework to deduce the package name automatically?

Should each test suite have a name (given as a parameter to `InstallTest`/`InstallTestSuite`/...)?  If so, then there could be another function for retrieving an installed test suite, given its name.  With a function called "Install"-something, I would expect that something like this happens, rather than that it justs creates an object and returns it.

**Comment by oysteins** [comment]
Comparing this package to Python's unittest module (referenced in an earlier comment), I found that there seems to be some differences in terminology:
- In Python, a "test" is a function which contains some assertions.  This corresponds to a "test case" in this package.
- In Python, a "test case" is a class containing tests (and optionally setup/teardown functions).  This corresponds to a "test suite" in this package.
- In Python, a "test suite" is a bundle of test cases and/or test suites.  There is nothing in this package which corresponds to that.

To agree with Python's terminology and feature set (which I assume is fairly standard), I suggest to do the following:
- Rename "test case" to "test".
- Rename "test suite" to "test case".
- Add a new kind of object called "test suite", which is used to combine test cases and/or test suites.

Combining this with other suggestions from comments above, I have tried to rewrite the examples in the package using a hypothetical new syntax:

``` GAP
additive := TestCase( function( Setup, TearDown, Test, FailingTest, a, b )
  # would be nice if "a+b" was available as both string and evaluated value

  Test("plus", function()
    AssertEqual( a, 1 * a );
    AssertEqual( a + a, 2 * a );
    # Commutative, associative
    AssertEqual( a + b, b + a );
    AssertEqual( (a + b) + a, a + (b + a) );
  end);

  Test("ainv", function()
    AssertEqual( -(-a), a );
    AssertEqual( -(a+b), -a - b );
    AssertEqual( -a, (-1) * a );
    AssertTrue( IsZero( a + (-a) ) );
  end);

  Test("zero", function()
    # is zero neutral and idempotent?
    AssertEqual( a + Zero(a), a );
    AssertEqual( Zero(a) + Zero(a), Zero(a) );
    AssertEqual( Zero(Zero(a)), Zero(a) );
    # Some alternative ways to produce zero
    AssertEqual( Zero( a ), a - a );
    AssertEqual( Zero( a ), 0 * a );
  end );

end );

additive_suite := TestSuite( [ TestCaseInstance( additive, 42, -19 ),
                               TestCaseInstance( additive, [ 17, 0 ], [ -17, 42 ] ) ] );

smallgroup := TestCase( function( Setup, TearDown, Test, FailingTest )
  local G, oldlevel;

  # Optionally perform some setup whenever an instance
  # of this test suite is run.
  Setup(function()
    oldlevel := InfoLevel(InfoWarning);
    SetInfoLevel(InfoWarning, 1);
    G := SmallGroup(8, 4);
    return true; # return false to indicate a setup failure
  end);

  # Optionally perform some setup after an instance
  # of this test suite has been run.
  TearDown(function()
    SetInfoLevel(InfoWarning, oldlevel);
  end);

  # a working test
  Test("group size", function()
    AssertEqual( Size(G), 8 );
  end);

  # another working test
  Test("small group id", function()
    AssertEqual( IdSmallGroup(G), [8, 4] );
  end);

  # a failing test (oops)
  Test("small group id", function()
    AssertTrue( IsAbelian(G) );
  end);

  # a known-to-fail test
  FailingTest("size_expected_fail", function()
    AssertEqual( Size(G), 7 );
  end);

  # a known-to-fail test that actually passes
  FailingTest("dummy_unexpected_pass", function()
    AssertEqual( 17, 17 );
  end);
end );

all_tests := TestSuite( [ additive_suite, smallgroup ] );

Run( all_tests );
```

Some features of this suggested syntax:
- `TestCase` takes a function as argument, creates a new test case object, and returns it.  The function that is given to `TestCase` should take at least four arguments `Setup`, `TearDown`, `Test` and `FailingTest`, and may take more arguments if it wants to be parametrized.
- Inside the function we supply to `TestCase`, we can call the four functions `Setup`, `TearDown`, `Test` and `FailingTest` to add stuff to the test case.
- We create an instance of a test case by calling `TestCaseInstance` with the test case object and any parameters we want to pass to the test case.  If the test case does not take any parameters, we can use the test case object itself as an instance.
- We create a test suite by calling `TestSuite` on a list of test cases and/or test suites.
- We run all tests in a suite by calling `Run` on the suite.  It should probably also be possible to call `Run` on a test case instance.

**Comment by fingolfin** [comment]
I really like this. Implementing it shouldn't be too hard. Some detals still would need to be hashed out, but that's normal.

As to `would be nice if "a+b" was available as both string and evaluated value` -- yeah, it would be, but I don't see how, without resorting to nasty tricks like passing in strings and using `Eval` on them. I am not aware of nice solutions for this in Python or Java (doesn't mean they don't exist, just that I am not aware). Of course in languages where more powerful syntax extensions (e.g. via hygenic macros) are possible, such as Lisp or Julia, this would be possible. Perhaps also in Ruby (?).

Anyway, this is of course why we have things like `AssertEqual` instead of just `AssertTrue`.  Perhaps the assertion in each test could also take a count, so that if one is triggered, it would print something like: `Assertion 3 in test "small group id" failed`.

BTW, I noticed you had two tests with the same name "small group id" -- was that intentional? I guess it should trigger an error.

**Comment by oysteins** [comment]
I made this example by rewriting the examples that already exist in the package's `examples` directory using my suggested syntax.  The comment about "a+b" being available as a string was already there, I just didn't remove it.  Let's see who wrote that:

```
~/src/UnitTests/examples$ git blame additive.unit | grep 'would be nice'
0b1b11ae (Max Horn 2013-01-10 01:22:40 +0100  6) # would be nice if "a+b" was available as both string and evaluated value
```

Anyway, I agree that it would be nice to have the expressions in assertions available as strings, but also that it's probably not worth it to use nasty tricks to accomplish this.

The duplicate test name also comes from your example files, so I can't answer whether it was intentional (and I also didn't notice the duplicate name while rewriting the example).

**Comment by fingolfin** [comment]
Hehe fair enough, I didn't notice this was just a rewrite (clearly it's been too long since I looked at this). Anyway, if the duplicate is due to me, then it certainly was an accident. But I asked to make sure I didn't miss some idea. E.g. one could concevable "chain" tests with identical names... though I think I'd prefer a simple error a la "test names must be unique, but name FOO was used before" or so.
